### DISCUSSION TITLE
High level algorithm for scaling down a PDF

### DISCUSSION THREAD

**Discussion by venkatd** [qpdf/qpdf#1209] [Q&A] [answered]
We sometimes get PDF uploads in our app that confuse our viewers because they have really large mediabox sizes.

If I use the qpdf json command, what do I need to do in addition to updating the mediabox and cropbox if I want to scale the content too without reducing the quality of anything? Is this a complex transformation?

Ideally minimal changes to the actual PDF structure!

I found something related in this javascript library but I'm not 100% sure what is going on:
https://github.com/Hopding/pdf-lib/issues/128#issuecomment-951606295

**Comment by jberkenbilt** [comment, marked as answer]
With PDF 1.6 and newer (and PDF 1.6 is OLD so that should be everything these days), this can be as simple as putting `/UserUnit` in the page dictionary. For example, `/UserUnit 0.5` will change the basic unit size to half its default value, effectively scaling the page by half without changing any of the boxes. The code you sited looks like its doing something different from that. I'm not familiar with the library in question, but my guess is that it's modifying the content stream to apply transformation matrices. This is another way to do it. You could change the bounding boxes (MediaBox, etc.) and then wrap the content with transformation operations. I'd try /UserUnit first through. Calculate a scale ratio by figuring out the maximum number you can multiply width and height by to keep them both below the desired value.

Example: if you wanted to fit within 612x792 (US Letter in Points, which are 1/72") and you had a MediaBox of 2000x2000:
* 612/2000 = 0.306
* 792/2000 = 0.396
* min(0.306, 0.396, 1.0) = 0.306  (throwing 1.0 in there to prevent this algorithm from _increasing_ the page size)
If you set /UserUnit to 0.306, the mediabox of 2000x2000 would translate to a page that is 612x612 points, which would fit within the US Letter page. Your ratio would be different depending on what goal you were specifically trying to achieve. You can use qpdf JSON to just add `/UserUnit 0.306` to the page dictionary.

**Comment by venkatd** [comment]
Thanks for the clear explanation! I will give this a try.

I wanted to say thank you for all the effort in this library. The json format v2 from qpdf has been a godsend. I've worked with libraries that over-abstract the PDF format which ended up doing more harm than good.

Working with an intermediate representation hits the sweet spot because I can do exactly what is needed without managing the very low level edge cases the PDF format.

**Reply by jberkenbilt** [reply]
Thanks for your kind words about JSON v2. It seems you have fully grasped its reason for being. I've always been a fan of light abstraction -- I like abstractions that take care of tedious, error-prone things but not abstractions that try to isolate you from technical details. qpdf tries to sit in that space.
